modules/platforms/dotnet/Apache.Ignite.Core.Tests/Client/Datastream/DataStreamerClientTest.cs

```diff
@@ -0,0 +1,831 @@
+            {
+                var ex = Assert.Throws<IgniteClientException>(() => streamer.Remove(1));
+
+                Assert.AreEqual("DataStreamer can't remove data when AllowOverwrite is false.", ex.Message);
+            }
+        }
+
+        /// <summary>
+        /// Tests streaming of relatively long list of entries to verify multiple buffer flush correctness.
+        /// </summary>
+        [Test]
+        [Category(TestUtils.CategoryIntensive)]
+        public void TestStreamLongList()
+        {
+            var cache = GetClientCache<int>();
+            const int count = 50000;
+
+            using (var streamer = Client.GetDataStreamer<int, int>(cache.Name))
+            {
+                for (var k = 0; k < count; k++)
+                {
+                    streamer.Add(k, -k);
+                }
+            }
+
+            Assert.AreEqual(count, cache.GetSize());
+            Assert.AreEqual(-2, cache[2]);
+            Assert.AreEqual(-200, cache[200]);
+        }
+
+        /// <summary>
+        /// Tests streamer usage from multiple threads.
+        /// </summary>
+        [Test]
+        [Category(TestUtils.CategoryIntensive)]
+        public void TestStreamMultithreaded()
+        {
+            var cache = GetClientCache<int>();
+            const int count = 250000;
+            int id = 0;
+
+            using (var streamer = Client.GetDataStreamer<int, int>(cache.Name))
+            {
+                TestUtils.RunMultiThreaded(() =>
+                {
+                    while (true)
+                    {
+                        var key = Interlocked.Increment(ref id);
+
+                        if (key > count)
+                        {
+                            break;
+                        }
+
+                        // ReSharper disable once AccessToDisposedClosure
+                        streamer.Add(key, key + 2);
+                    }
+                }, 8);
+            }
+
+            Assert.AreEqual(count, cache.GetSize());
+            Assert.AreEqual(4, cache[2]);
+            Assert.AreEqual(22, cache[20]);
+        }
+
+        /// <summary>
+        /// Tests streamer usage with Parallel.For, which dynamically allocates threads to perform work.
+        /// This test verifies backpressure behavior quite well.
+        /// </summary>
+        [Test]
+        [Category(TestUtils.CategoryIntensive)]
+        public void TestStreamParallelFor()
+        {
+            var cache = GetClientCache<int>();
+            const int count = 250000;
+
+            using (var streamer = Client.GetDataStreamer<int, int>(cache.Name))
+            {
+                // ReSharper disable once AccessToDisposedClosure
+                Parallel.For(0, count, i => streamer.Add(i, i + 2));
+
+                streamer.Flush();
+                CheckArrayPoolLeak(streamer);
+            }
+
+            var size = cache.GetSize();
+            if (size != count)
```

Comment:
We can lose data due to race conditions and other concurrency bugs, like buffer pooling/sharing issues.

Reply:
Still didn't get it. Why we can't get such races and bugs in `TestStreamMultithreaded`? As far as I understand this test does almost the same. Also, can we lose keys `4` and `22` (checked after size check)?

Reply:
`Parallel.For` dynamically (and rather aggressively) assigns threads depending on the load behavior and machine configuration. This test proved to be useful during development because it stressed the system a lot more and highlighted some performance issues. We can remove `TestStreamMultithreaded` instead if you insist.